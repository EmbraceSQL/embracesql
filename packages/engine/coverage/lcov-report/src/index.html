
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> src</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">68.84% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>137/199</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">48.28% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>14/29</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">60% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>30/50</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">69.43% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>134/193</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="configuration.ts"><a href="configuration.ts.html">configuration.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="13" class="abs high">13/13</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="13" class="abs high">13/13</td>
	</tr>

<tr>
	<td class="file low" data-value="context.ts"><a href="context.ts.html">context.ts</a></td>
	<td data-value="42.86" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 42%"></div><div class="cover-empty" style="width: 58%"></div></div>
	</td>
	<td data-value="42.86" class="pct low">42.86%</td>
	<td data-value="7" class="abs low">3/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="42.86" class="pct low">42.86%</td>
	<td data-value="7" class="abs low">3/7</td>
	</tr>

<tr>
	<td class="file high" data-value="generator.ts"><a href="generator.ts.html">generator.ts</a></td>
	<td data-value="81.25" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 81%"></div><div class="cover-empty" style="width: 19%"></div></div>
	</td>
	<td data-value="81.25" class="pct high">81.25%</td>
	<td data-value="16" class="abs high">13/16</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="9" class="abs medium">6/9</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="86.67" class="pct high">86.67%</td>
	<td data-value="15" class="abs high">13/15</td>
	</tr>

<tr>
	<td class="file low" data-value="inprocess.ts"><a href="inprocess.ts.html">inprocess.ts</a></td>
	<td data-value="22.22" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 22%"></div><div class="cover-empty" style="width: 78%"></div></div>
	</td>
	<td data-value="22.22" class="pct low">22.22%</td>
	<td data-value="9" class="abs low">2/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="20" class="pct low">20%</td>
	<td data-value="5" class="abs low">1/5</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="8" class="abs low">2/8</td>
	</tr>

<tr>
	<td class="file medium" data-value="render.ts"><a href="render.ts.html">render.ts</a></td>
	<td data-value="67.21" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 67%"></div><div class="cover-empty" style="width: 33%"></div></div>
	</td>
	<td data-value="67.21" class="pct medium">67.21%</td>
	<td data-value="61" class="abs medium">41/61</td>
	<td data-value="38.89" class="pct low">38.89%</td>
	<td data-value="18" class="abs low">7/18</td>
	<td data-value="68.42" class="pct medium">68.42%</td>
	<td data-value="19" class="abs medium">13/19</td>
	<td data-value="66.1" class="pct medium">66.1%</td>
	<td data-value="59" class="abs medium">39/59</td>
	</tr>

<tr>
	<td class="file low" data-value="server.ts"><a href="server.ts.html">server.ts</a></td>
	<td data-value="22.58" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 22%"></div><div class="cover-empty" style="width: 78%"></div></div>
	</td>
	<td data-value="22.58" class="pct low">22.58%</td>
	<td data-value="31" class="abs low">7/31</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="23.33" class="pct low">23.33%</td>
	<td data-value="30" class="abs low">7/30</td>
	</tr>

<tr>
	<td class="file high" data-value="structured-console.ts"><a href="structured-console.ts.html">structured-console.ts</a></td>
	<td data-value="95.74" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 95%"></div><div class="cover-empty" style="width: 5%"></div></div>
	</td>
	<td data-value="95.74" class="pct high">95.74%</td>
	<td data-value="47" class="abs high">45/47</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="88.89" class="pct high">88.89%</td>
	<td data-value="9" class="abs high">8/9</td>
	<td data-value="95.65" class="pct high">95.65%</td>
	<td data-value="46" class="abs high">44/46</td>
	</tr>

<tr>
	<td class="file high" data-value="watcher.ts"><a href="watcher.ts.html">watcher.ts</a></td>
	<td data-value="86.67" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 86%"></div><div class="cover-empty" style="width: 14%"></div></div>
	</td>
	<td data-value="86.67" class="pct high">86.67%</td>
	<td data-value="15" class="abs high">13/15</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="86.67" class="pct high">86.67%</td>
	<td data-value="15" class="abs high">13/15</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Sun May 31 2020 10:35:53 GMT-0500 (Central Daylight Time)
            </div>
        </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    