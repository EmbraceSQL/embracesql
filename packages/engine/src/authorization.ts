import { HasHeaders, HasToken, HasConfiguration } from "./shared-context";
import { parseBearerToken, FileCache, validate } from "@embracesql/identity";
import path from "path";

/**
 * Built in macro to validate OpenID tokens.
 */
export const validateOpenIDAuthorization = async (
  context: HasHeaders & HasToken & HasConfiguration
): Promise<void> => {
  // we will eat errors, validation isn't mandatory.
  try {
    context.token = await validate(parseBearerToken(context.headers), {
      cache: FileCache(
        path.resolve(context.configuration.embraceSQLRoot, ".embracesql", "jwk")
      ),
      ignoreExp: context.configuration.ignoreExp,
    });
    return;
  } catch (e) {
    console.assert(e);
    //TODO -- error event
  }
};
