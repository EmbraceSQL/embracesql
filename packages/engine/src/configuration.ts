import { cosmiconfig } from "cosmiconfig";
import { generateFromTemplates } from "./generator";
import path from "path";
import { Configuration } from "./shared-context";

/**
 * Load up a configuration object.
 *
 * This will interpolate environment variables, build and object and hand it back.
 *
 * Will look in the current directory or a environment variable set root.
 */
export const loadConfiguration = async (
  root: string
): Promise<Configuration> => {
  // default name based on the path
  const name = root.split(path.sep).slice().pop() || "default";
  // run the root generation templates, gives you something to work
  // with even if you start in an empty directory so that system 'always works'
  await generateFromTemplates(
    {
      configuration: {
        name,
        embraceSQLRoot: root,
      },
      databases: undefined,
      directQueryExecutors: {},
      autocrudExecutors: {},
      close: (): Promise<void> => {
        return;
      },
    },
    "default"
  );
  // TODO env var substition loader hook
  // going with cosmic config -- even though this is just doing YAML for the moment
  const explorer = cosmiconfig("embracesql", {
    searchPlaces: ["embracesql.yaml", "embracesql.yml"],
  });
  const result = await explorer.search(root);
  const config = result.config as Configuration;
  config.embraceSQLRoot = root;
  return {
    ...config,
  };
};
