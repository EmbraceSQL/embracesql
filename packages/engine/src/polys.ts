/* eslint-disable @typescript-eslint/ban-types */
import { CanBatchSet } from "./shared-context";

/**
 * Not sure this is the best name -- but I'm calling them Polys.
 *
 * What's a poly -- it's an array or a scalar, with some convenience
 * to be able to scalar set 'all' of the array member, or get the first
 * member as is a common case when you want to ignore the fact that
 * parameters can be an array/batch.
 *
 */
export const polyArray = <T extends object, K>(
  target: Iterable<T>
): CanBatchSet<T> => {
  const spreader = {
    // setting sets them all
    set: (target: Iterable<T>, prop: PropertyKey, value: K): boolean => {
      [...target].forEach((t) => (t[prop] = value));
      return true;
    },
    // getting gets the first item if we can
    // the assumption is you will often have parameter 'sets' of one
    get: (target: Iterable<T>, prop: PropertyKey): K => {
      // -- get at array members, like .length
      if (prop in target) return target[prop];
      // otherwise peek into the elements of the array and read the first
      const buffer = [...target];
      return buffer.length ? buffer[0][prop] : undefined;
    },
  };
  return (new Proxy<Iterable<T>>(target, spreader) as unknown) as CanBatchSet<
    T
  >;
};
