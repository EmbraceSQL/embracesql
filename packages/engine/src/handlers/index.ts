import { InternalContext } from "../internal-context";
import sqlModulePipeline from "./sqlmodule-pipeline";
import contextPipeline from "./context-pipeline";
import autoCrudPipeline from "./autocrud-pipeline";
import generateDefaultHandlers from "./generate-default-handlers";

/**
 * Scrub up identifiers to be valid JavaScript names.
 */
export const identifier = (key: string): string => {
  const id = key
    .trim()
    // quotes won't do
    .replace(/['"]+/g, "")
    // snake case
    .replace(/\W+/g, "_");
  return /^\d/.test(id) ? "_" + id : id;
};

/**
 * Manage a directory of event handlers, which form the basis of APIs
 *
 * All SQL files in the directory serve as entry points. The process is to enumerate
 * all of these SQL files and build an in memory file system analog with hashes
 * to be used as cache keys.
 *
 * * Loads up and hashes all sql files found
 * * Creates a database specific wrapper for invoking each SQL.
 * * Makes sure all the needed event handlers are generated
 *
 * @param rootContext - like other internal methods, run off the root context
 */
export const embraceEventHandlers = async (
  rootContext: InternalContext
): Promise<InternalContext> => {
  // all the on disk sql modules and their handlers
  await sqlModulePipeline(rootContext);
  // auto auto CRUD ... I wannt be ... an Auto CRUD
  await autoCrudPipeline(rootContext);
  // every module is now generated -- time for handler generation
  for (const database of Object.values(rootContext.databases)) {
    for (const module of Object.values(database.modules)) {
      await generateDefaultHandlers(rootContext, module);
    }
  }
  // stitch together the full context and final combined files
  // this is the 'pack' phase
  await contextPipeline(rootContext);
  return rootContext;
};
