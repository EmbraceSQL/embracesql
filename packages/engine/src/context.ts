import { embraceDatabases } from "./database-engines";
import { embraceEventHandlers } from "./handlers";
import {
  Database,
  SQLModule,
  SQLColumnMetadata,
  SQLParameters,
  SQLRow,
  HasConfiguration,
  Configuration,
  SQLTableMetadata,
  AutocrudModule,
  SQLModuleExecutor,
  AutocrudExecutor,
} from "./shared-context";
import { AST } from "node-sql-parser";
import { SQLModuleInternal } from "./handlers/sqlmodule-pipeline";

/**
 * Keep track of individual migration files with this type.
 */
export type MigrationFile = {
  name: string;
  content: string;
};

/**
 * This is the tree of paths derived from SQL files on disk.
 * This is in a compressed path format, so each key can have / in it.
 */
export type SQLModules = {
  [index: string]: SQLModule;
};

/**
 * All available autocrud modules.
 * This is in a compressed path format, so each key can have / in it.
 */
export type AutocrudModules = {
  [index: string]: AutocrudModule;
};

/**
 * SQL Pair for C in CRUD.
 */
export type CreateAndReadbackSQL = {
  create: string;
  readback: string;
};

/**
 * SQL for the R in CRUD.
 */
export type ReadSQL = {
  allRows: string;
  byKey: string;
};

/**
 * A single instance of a database for use internally.
 */
export type DatabaseInternal = Database & {
  /**
   * Execute the sql module query on this database, and
   * promise some result.
   *
   * @param sql - sql string in the dialect of the target database, can include :name style paramters
   * @param parameters - name value pairs are the passed parameters
   */
  execute: (sql: string, parameters?: SQLParameters) => Promise<SQLRow[]>;
  /**
   * Analyze the passed module and determine the resultset type(s).
   */
  analyze: (sqlModule: SQLModuleInternal) => Promise<SQLColumnMetadata[]>;
  /**
   * Parse out the SQL.
   */
  parse: (SQLModule) => AST[] | AST;
  /**
   * Do a migration.
   */
  migrate: (migrationFile: MigrationFile) => Promise<void>;
  /**
   * Clean close.
   */
  close: () => Promise<void>;
  /**
   * Get schematic information on all tables.
   */
  schema: () => Promise<SQLTableMetadata[]>;
  /**
   * Get the full create statement sql, with an immeidate readback
   */
  createSQL: (sqlTable: SQLTableMetadata) => Promise<CreateAndReadbackSQL>;
  /**
   * Get the SQL to read back a table.
   */
  readSQL: (sqlTable: SQLTableMetadata) => Promise<ReadSQL>;
};

/**
 * Collections of modules.
 */
export type HasModules = {
  /**
   * All modules for this database.
   */
  sqlModules: SQLModules;
  /**
   * Autocrud for this database.
   */
  autocrudModules: AutocrudModules;
};

/**
 * Fully wrapped internal database representation.
 */
export type DatabaseInternalWithModules = DatabaseInternal & HasModules;

/**
 * All the databases from the internal point of view.
 */
export type AllDatabasesInternal = {
  [index: string]: DatabaseInternalWithModules;
};

/**
 * The root context is the context of configuration, databases, and sql modules and is
 * used to drive code generation and runtime execution 'in the engine' of EmbraceSQL
 *
 * The type here is a bit different from the context used in handlers, it has more metadata!
 */
export type InternalContext = HasConfiguration & {
  /**
   * All configured databases, by name.
   */
  databases: AllDatabasesInternal;
  /**
   * Close all DB connections.
   */
  close: () => Promise<void>;
  /**
   * Ability to execute sql modules.
   */
  sqlModuleExecutors: {
    [index: string]: SQLModuleExecutor;
  };
  /**
   * Ability to execute autocrud modules.
   */
  autocrudModuleExecutors: {
    [index: string]: AutocrudExecutor;
  };
};

/**
 * With a configuration in hand, set up a new rootContext.
 *
 * This is built to be called -- repeatedly if needed. The idea is you can watch, and
 * rebuild a whole new context as needed -- swapping the root context at runtime to
 * hot-reconfigure the system without worrying about any state leaking.
 *
 * @param configuration - build a root context from this configuration.
 */
export const buildInternalContext = async (
  configuration: Configuration
): Promise<InternalContext> => {
  const databases = await embraceDatabases(configuration);
  const internalContext = {
    configuration,
    databases,
    sqlModuleExecutors: {},
    autocrudModuleExecutors: {},
    close: async (): Promise<void> => {
      const waitForThem = Object.values(databases).map((database) =>
        database.close()
      );
      await Promise.all(waitForThem);
    },
  };
  // need the database first, their connections are used
  // to mine metadata
  await embraceDatabases(configuration);
  await embraceEventHandlers(internalContext);
  return internalContext;
};
