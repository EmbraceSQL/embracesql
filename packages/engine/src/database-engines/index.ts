import embraceSQLite from "./sqlite";
import { AllDatabasesInternal } from "../context";
import { DatabaseInternal } from "../context";
import pLimit from "p-limit";
import {
  SQLColumnMetadata,
  SQLRow,
  SQLParameters,
  Configuration,
  CommonDatabaseModule,
  SQLParameterSet,
  SQLTableMetadata,
} from "../shared-context";
import { SQLModuleInternal } from "../handlers/sqlmodule-pipeline";
import Url from "url-parse";
import graphlib from "graphlib";

/**
 * Serialize database use per process as we really only have one connection.
 */
const oneAtATime = pLimit(1);

/**
 * Filter to just the declared parameters. This lets callers be a little sloppy
 * and things will still 'just work'.
 */
export const validParameters = (
  module: CommonDatabaseModule,
  parameters: SQLParameterSet
): SQLParameterSet =>
  Object.fromEntries(
    module.namedParameters.map((p) => [p.name, parameters[p.name]])
  );

/**
 * Embrace a database, bringing it into context.
 *
 * TODO: this needs a wrapper around each database to try to heal itself
 * and reconnect, or to fault and let the container worker process end.
 */
const embraceSingleDatabase = async (
  configuration: Configuration,
  databaseName: string
): Promise<DatabaseInternal> => {
  const dbUrl = new Url(configuration.databases[databaseName]);
  switch (dbUrl.protocol.split(":")[0].toLowerCase()) {
    case "sqlite":
      return embraceSQLite(configuration, databaseName, dbUrl);
    default:
      return undefined;
  }
};

/**
 * Referential graph. This is undirected, relationships are considered
 * symmetric even though the SQL references clause it actually directed.
 *
 * Reason being -- 1-man-1 jointer tables, those joiners 'reference' the parent, compared
 * to lookup value type tables where the parent references the child.
 *
 * And -- master detail 1-many, the child references the parent, but you will
 * want to get all the children along with a fetch of a parent to build screens
 * in actual practice.
 */
export const buildReferentialGraph = async (
  tables: SQLTableMetadata[]
): Promise<void> => {
  // each table is a node, each reference is an undirected edge
  const graph = new graphlib.Graph({ multigraph: true });
  for (const table of tables) {
    graph.setNode(`${table.schema}.${table.name}`, table);
    for (const reference of table.references) {
      graph.setEdge(
        `${table.schema}.${table.name}`,
        `${reference.toSchema}.${reference.toTable}`,
        reference
      );
    }
  }
};

/**
 * Every database in the configuraton is embraced and brought into the context.
 */
export const embraceDatabases = async (
  configuration: Configuration
): Promise<AllDatabasesInternal> => {
  // name value pairs inside the root context -- there isn't a strongly
  // typed generated context type available to the generator itself
  const databases = {};
  Object.keys(configuration.databases).forEach(async (databaseName) => {
    const database = await embraceSingleDatabase(configuration, databaseName);
    databases[databaseName] = {
      ...database,
      autocrudModules: {},
      sqlModules: {},
      // here is wrapping the individual database driver execute and analyze
      // with a throttled promise limit -- since we have only one connection
      // these are not re-entrant
      execute: async (
        sql: string,
        parameters?: SQLParameters
      ): Promise<SQLRow[]> => {
        return oneAtATime(() => database.execute(sql, parameters));
      },
      analyze: async (
        sqlModule: SQLModuleInternal
      ): Promise<SQLColumnMetadata[]> => {
        return oneAtATime(() => database.analyze(sqlModule));
      },
    };
  });
  return databases;
};
