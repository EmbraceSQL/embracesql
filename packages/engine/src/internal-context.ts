import { embraceDatabases } from "./database-engines";
import { embraceEventHandlers } from "./handlers";
import {
  Database,
  SQLModule,
  SQLColumnMetadata,
  SQLRow,
  HasConfiguration,
  Configuration,
  AutocrudModule,
  Closeable,
  EntryPoint,
  SQLParameterSet,
} from "./shared-context";
import { AST } from "node-sql-parser";
import { SQLModuleInternal } from "./handlers/sqlmodule-pipeline";

/**
 * Keep track of individual migration files with this type.
 */
export type MigrationFile = {
  name: string;
  content: string;
};

/**
 * This is the tree of paths derived from SQL files on disk.
 * This is in a compressed path format, so each key can have / in it.
 */
export type Modules = {
  [index: string]: SQLModule | AutocrudModule;
};

/**
 * SQL Pair for C in CRUD.
 */
export type CreateAndReadbackSQL = {
  create: string;
  readback: string;
};

/**
 * SQL for the R in CRUD.
 */
export type ReadSQL = {
  allRows: string;
  byKey: string;
};

/**
 * SQL for the U in CRUD.
 */
export type UpdateSQL = {
  byKey: string;
};

/**
 * Function wrapper for atomicicity in running in the database.
 */
export interface AtomicDatabaseAction {
  /**
	@param fn - Promise-returning/async function.
	@returns The promise returned by calling `fn(...arguments)`.
	*/
  <ReturnType>(fn: () => PromiseLike<ReturnType> | ReturnType): Promise<
    ReturnType
  >;
}

/**
 * A single instance of a database for use internally.
 */
export type DatabaseInternal = Database & {
  /**
   * Execute the sql module query on this database, and
   * promise some result.
   *
   * @param sql - sql string in the dialect of the target database, can include :name style paramters
   * @param parameters - name value pairs are the passed parameters
   */
  execute: (sql: string, parameters?: SQLParameterSet) => Promise<SQLRow[]>;
  /**
   * Analyze the passed module and determine the resultset type(s).
   */
  analyze: (sqlModule: SQLModuleInternal) => Promise<SQLColumnMetadata[]>;
  /**
   * Parse out the SQL.
   */
  parse: (SQLModule) => AST[] | AST;
  /**
   * Do a migration.
   */
  migrate: (migrationFile: MigrationFile) => Promise<void>;
  /**
   * Clean close.
   */
  close: () => Promise<void>;
  /**
   * Get the full create statement sql, with an immeidate readback
   */
  createSQL: (autocrudModule: AutocrudModule) => Promise<CreateAndReadbackSQL>;
  /**
   * Get the SQL to read back a table.
   */
  readSQL: (autocrudModule: AutocrudModule) => Promise<ReadSQL>;
  /**
   * Get the SQL to update a table.
   */
  updateSQL: (autocrudModule: AutocrudModule) => Promise<UpdateSQL>;
  /**
   * Throttle access to a database for atomicity. This is needed when thre
   * are multiple parallel queries possible and you need to batch -- such
   * as reading back a created key in the face of parallel inserts.
   */
  atomic: AtomicDatabaseAction;
};

/**
 * Collections of modules.
 */
export type HasModules = {
  /**
   * All modules for this database.
   */
  modules: Modules;
};

/**
 * Fully wrapped internal database representation.
 */
export type DatabaseInternalWithModules = DatabaseInternal & HasModules;

/**
 * All the databases from the internal point of view.
 */
export type AllDatabasesInternal = {
  [index: string]: DatabaseInternalWithModules;
};

/**
 * The root context is the context of configuration, databases, and sql modules and is
 * used to drive code generation and runtime execution 'in the engine' of EmbraceSQL
 *
 * The type here is a bit different from the context used in handlers, it has more metadata!
 */
export type InternalContext = HasConfiguration &
  Closeable & {
    /**
     * All configured databases, by name.
     */
    databases: AllDatabasesInternal;
    /**
     * Ability to execute sql modules.
     */
    moduleExecutors: {
      [index: string]: EntryPoint;
    };
  };

/**
 * With a configuration in hand, set up a new rootContext.
 *
 * This is built to be called -- repeatedly if needed. The idea is you can watch, and
 * rebuild a whole new context as needed -- swapping the root context at runtime to
 * hot-reconfigure the system without worrying about any state leaking.
 *
 * @param configuration - build a root context from this configuration.
 */
export const buildInternalContext = async (
  configuration: Configuration
): Promise<InternalContext> => {
  const databases = await embraceDatabases(configuration);
  const internalContext = {
    configuration,
    databases,
    moduleExecutors: {},
    close: async (): Promise<void> => {
      const waitForThem = Object.values(databases).map((database) =>
        database.close()
      );
      await Promise.all(waitForThem);
    },
  };
  // need the database first, their connections are used
  // to mine metadata
  await embraceDatabases(configuration);
  await embraceEventHandlers(internalContext);
  return internalContext;
};
