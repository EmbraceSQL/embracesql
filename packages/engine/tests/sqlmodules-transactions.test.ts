import path from "path";
import fs from "fs-extra";
import { loadConfiguration } from "../src/configuration";
import { buildInternalContext, InternalContext } from "../src/context";
import { migrate } from "../src/migrations";
import rmfr from "rmfr";

describe("sqlmodules provide transactions", () => {
  let rootContext: InternalContext;
  const root = path.relative(process.cwd(), "./.tests/sqlmodules-transactions");
  beforeAll(async () => {
    // clean up
    await fs.ensureDir(root);
    await rmfr(root);
    await fs.copy(
      path.join(__dirname, "configs/sqlmodules-transactions"),
      root
    );
    // get the configuration and generate - let's do this just the once
    // and have a few tests that asser things happened
    const configuration = await loadConfiguration(root);
    rootContext = await buildInternalContext(configuration);
    await migrate(rootContext);
    // reset and go
    await rootContext.close();
    rootContext = await buildInternalContext(configuration);
  });
  it("writes, reads back, and reads again with a transaction", async () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { EmbraceSQLEmbedded } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot
    ));
    const engine = await EmbraceSQLEmbedded();
    // initial values
    const initial = await engine.databases.default.all();
    // add a thing
    expect(
      await engine.databases.default.add({ id: 3, name: "automobiles" })
    ).toMatchSnapshot();
    // and read again
    expect(await engine.databases.default.all()).toEqual(initial);
  });
});
