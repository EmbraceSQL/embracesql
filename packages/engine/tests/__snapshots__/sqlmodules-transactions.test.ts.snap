// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqlmodules provide transactions writes, reads back, and reads again with a transaction 1`] = `
Array [
  Object {
    "id": 1,
    "name": "planes",
  },
  Object {
    "id": 2,
    "name": "trains",
  },
  Object {
    "id": 3,
    "name": "automobiles-ahem",
  },
]
`;
